src/libcrun/linux.c

```diff
@@ -1815,19 +1816,48 @@ create_missing_devs (libcrun_container_t *container, bool binds, libcrun_error_t
 static int
 do_masked_and_readonly_paths (libcrun_container_t *container, libcrun_error_t *err)
 {
-  size_t i;
-  int ret;
+  size_t i, j;
+  char *real_path;
+  int ret, mount_fd;
+  unsigned long flags = 0;
   runtime_spec_schema_config_schema *def = container->container_def;
+  proc_fd_path_t proc_buf;
+  struct statfs sfs;
 
   for (i = 0; i < def->linux->masked_paths_len; i++)
     {
-      ret = do_masked_or_readonly_path (container, def->linux->masked_paths[i], false, err);
+      ret = do_masked_or_readonly_path (container, def->linux->masked_paths[i], false, 0, err);
       if (UNLIKELY (ret < 0))
         return ret;
     }
   for (i = 0; i < def->linux->readonly_paths_len; i++)
     {
-      ret = do_masked_or_readonly_path (container, def->linux->readonly_paths[i], true, err);
+      flags = 0;
+      // Before setting path to read only we must check if any existing mount is a parent
+      // if yes then we must inherit the flags/options from the parent.
+      for (j = 0; j < def->mounts_len; j++)
+        {
+          if (strcmp (def->linux->readonly_paths[i], def->mounts[j]->destination) == 0)
+            {
+              mount_fd = open_mount_target (container, def->linux->readonly_paths[i], err);
```

Comment:
we already open the mount target inside `do_masked_or_readonly_path`, so we can make the statfs there, no?

This means we need to change `do_masked_or_readonly_path` to accept a bool `keep_flags` instead of the additional `flags`

Reply:
Yes SGTM. I'll amend it one thing which i saw it inheriting flags from `/proc` blows up  with `EINVAL` i need to cancel some of the flags which is causing that.

Reply:
Done, also a test case for the `/proc` use-case.